src/host-config.ts

```diff
@@ -164,7 +192,22 @@ async function readHostname() {
 
 async function setHostname(val: string) {
 	await config.set({ hostname: val });
-	await dbus.restartService('resin-hostname');
+
+	try {
+		// restart balena-hostname if it is loaded and NOT PartOf config-json.target
+		if (
+			!(await dbus.servicePartOf('balena-hostname')).includes(
+				'config-json.target',
+			)
+		) {
+			await dbus.restartService('balena-hostname');
```

Comment:
To avoid the nested try/catch you could do 

```
await Bluebird.any([dbus.restartService('balena-hostname'), dbus.restartService('resin-hostname')])
```

Which will try both alternatives and only throw if both fail.

Unfortunately that is not implemented yet on standard node https://www.npmjs.com/package/promise.any